src/linear_algebra/cayley_hamilton.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2020 Scott Morrison. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Scott Morrison
+-/
+import ring_theory.matrix_algebra
+import ring_theory.polynomial_algebra
+import linear_algebra.nonsingular_inverse
+
+/-!
+# The Cayley-Hamilton theorem, over a commutative ring.
+
+We use a nice proof from http://drorbn.net/AcademicPensieve/2015-12/CayleyHamilton.pdf
+-/
+
+noncomputable theory
+
+universes u v w
+
+open polynomial matrix
+open_locale big_operators
+
+variables {R : Type u} [comm_ring R]
+variables {n : Type w} [fintype n] [decidable_eq n]
+
+open finset
+
+/--
+The "characteristic matrix" of `M : matrix n n R` is the matrix of polynomials $t I - M$.
+The determinant of this matrix is the characteristic polynomial.
+-/
+def char_matrix (M : matrix n n R) : matrix n n (polynomial R) :=
+matrix.scalar n (X : polynomial R) - (λ i j, C (M i j))
+
+@[simp] lemma char_matrix_apply_eq (M : matrix n n R) (i : n) :
```

Comment:
The similar-looking lemmas in data.matrix.basic are named with "val". I don't know which convention lines up better with the rest of mathlib.
```suggestion
@[simp] lemma char_matrix_val_eq (M : matrix n n R) (i : n) :
```

Reply:
I'm pretty sure `apply` is what's used everywhere else. I'll make a separate PR to fix `data.matrix.basic`.

Reply:
#3297

Reply:
I'm going to mark this as resolved --- it turns out we're already not consistent, and #3297 can be the "tracking issue" for sorting this out.